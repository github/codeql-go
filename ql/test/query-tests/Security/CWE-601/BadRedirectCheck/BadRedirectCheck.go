package main

import "net/http"

func sanitizeUrl(redir string) string {
	if len(redir) > 0 && redir[0] == '/' {
		return redir
	}
	return "/"
}

func badRedirect(redirect string, rw http.ResponseWriter, req *http.Request) {
	http.Redirect(rw, req, sanitizeUrl(redirect), 302)
}