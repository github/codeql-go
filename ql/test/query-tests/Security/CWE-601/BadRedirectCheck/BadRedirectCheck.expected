edges
| BadRedirectCheck.go:5:18:5:22 | argument corresponding to redir : string | BadRedirectCheck.go:7:10:7:14 | redir : string |
| BadRedirectCheck.go:7:10:7:14 | redir : string | BadRedirectCheck.go:13:25:13:45 | call to sanitizeUrl |
| BadRedirectCheck.go:12:18:12:25 | argument corresponding to redirect : string | BadRedirectCheck.go:13:37:13:44 | redirect : string |
| BadRedirectCheck.go:13:37:13:44 | redirect : string | BadRedirectCheck.go:13:25:13:45 | call to sanitizeUrl |
| cves.go:14:23:14:25 | argument corresponding to url : string | cves.go:16:26:16:28 | url |
| cves.go:33:14:33:34 | call to Get : string | cves.go:37:25:37:32 | redirect |
| cves.go:41:14:41:34 | call to Get : string | cves.go:45:25:45:32 | redirect |
| main.go:18:24:18:26 | argument corresponding to url : string | main.go:20:26:20:28 | url |
nodes
| BadRedirectCheck.go:5:18:5:22 | argument corresponding to redir : string | semmle.label | argument corresponding to redir : string |
| BadRedirectCheck.go:7:10:7:14 | redir : string | semmle.label | redir : string |
| BadRedirectCheck.go:12:18:12:25 | argument corresponding to redirect : string | semmle.label | argument corresponding to redirect : string |
| BadRedirectCheck.go:13:25:13:45 | call to sanitizeUrl | semmle.label | call to sanitizeUrl |
| BadRedirectCheck.go:13:37:13:44 | redirect : string | semmle.label | redirect : string |
| cves.go:14:23:14:25 | argument corresponding to url : string | semmle.label | argument corresponding to url : string |
| cves.go:16:26:16:28 | url | semmle.label | url |
| cves.go:33:14:33:34 | call to Get : string | semmle.label | call to Get : string |
| cves.go:37:25:37:32 | redirect | semmle.label | redirect |
| cves.go:41:14:41:34 | call to Get : string | semmle.label | call to Get : string |
| cves.go:45:25:45:32 | redirect | semmle.label | redirect |
| main.go:18:24:18:26 | argument corresponding to url : string | semmle.label | argument corresponding to url : string |
| main.go:20:26:20:28 | url | semmle.label | url |
#select
| BadRedirectCheck.go:6:23:6:37 | ...==... | BadRedirectCheck.go:5:18:5:22 | argument corresponding to redir : string | BadRedirectCheck.go:13:25:13:45 | call to sanitizeUrl | This is a check that $@, which flows into a $@, has a leading slash, but not that does not have '/' or '\\' in its second position. | BadRedirectCheck.go:5:18:5:22 | argument corresponding to redir | this value | BadRedirectCheck.go:13:25:13:45 | call to sanitizeUrl | redirect |
| BadRedirectCheck.go:13:25:13:45 | call to sanitizeUrl | BadRedirectCheck.go:12:18:12:25 | argument corresponding to redirect : string | BadRedirectCheck.go:13:25:13:45 | call to sanitizeUrl | This is a check that $@, which flows into a $@, has a leading slash, but not that does not have '/' or '\\' in its second position. | BadRedirectCheck.go:12:18:12:25 | argument corresponding to redirect | this value | BadRedirectCheck.go:13:25:13:45 | call to sanitizeUrl | redirect |
| cves.go:15:5:15:24 | call to isValidRedirect | cves.go:14:23:14:25 | argument corresponding to url : string | cves.go:16:26:16:28 | url | This is a check that $@, which flows into a $@, has a leading slash, but not that does not have '/' or '\\' in its second position. | cves.go:14:23:14:25 | argument corresponding to url | this value | cves.go:16:26:16:28 | url | redirect |
| cves.go:34:6:34:37 | call to HasPrefix | cves.go:33:14:33:34 | call to Get : string | cves.go:37:25:37:32 | redirect | This is a check that $@, which flows into a $@, has a leading slash, but not that does not have '/' or '\\' in its second position. | cves.go:33:14:33:34 | call to Get | this value | cves.go:37:25:37:32 | redirect | redirect |
| cves.go:42:6:42:37 | call to HasPrefix | cves.go:41:14:41:34 | call to Get : string | cves.go:45:25:45:32 | redirect | This is a check that $@, which flows into a $@, has a leading slash, but not that does not have '/' or '\\' in its second position. | cves.go:41:14:41:34 | call to Get | this value | cves.go:45:25:45:32 | redirect | redirect |
| main.go:19:5:19:21 | call to isValidRedir | main.go:18:24:18:26 | argument corresponding to url : string | main.go:20:26:20:28 | url | This is a check that $@, which flows into a $@, has a leading slash, but not that does not have '/' or '\\' in its second position. | main.go:18:24:18:26 | argument corresponding to url | this value | main.go:20:26:20:28 | url | redirect |
